travis: add legacy macos build and custom Qt version

dyfer opened the pull request:
<!-- Please see CONTRIBUTING.md for guidelines. -->

## Purpose and Motivation

~~EDIT: this PR will include using custom Qt version in regular builds to fix #5167 and #5175~~

Inspired by [this comment](https://github.com/supercollider/supercollider/issues/5033#issuecomment-696100369) and following conversation, I wanted to see if adding a "legacy" build to our build matrix would be an option.

In this PR I propose to add one more macOS build to our travis build matrix, based on XCode 10 running on macOS 13 and using Qt 5.9. IIUC, this build should run on systems as old as 10.10 (untested). 

I propose this build to be an **unsupported** legacy build, as a courtesy to users running older systems. Once maintaining this becomes a burden, we'd just drop it. But for now it seems to build just fine. 

~~As far as implementation goes, I ended up using a modified QT formula on my own github, I don't see a better way. I understand this might be too hacky of an approach for an addition to this codebase.~~ Now using formula in `supercollider/formulae` repo.

~~TODO: add uploading of the build to AWS~~ Added deployment to Github and AWS


## Types of changes

<!-- Delete lines that don't apply -->

- New feature (?)

## To-do list

<!-- Complete an item by checking it: [x]. Add new entries to track your progress -->

- [x] Deployment to Github/AWS is added
- [x] Code is tested
- [x] This PR is ready for review

brianlheim commented on .travis.yml:
```diff
@@ -235,6 +235,17 @@ matrix:
+      env:
+        - JOB_NAME="macos xcode10.1 legacy build with qt 5.9"
+        - QT=true
+        - QT_FORMULA="dyfer/qt/qt@5.9.3"
```
where is this formula maintained? would it be easy to have an sc-org fork instead?

dyfer replied:
I hacked this formula together, it's in [my repo](https://github.com/dyfer/homebrew-qt). Totally unofficial :) 
I'd be happy to transfer it to sc-org, but wasn't sure if we want that.

brianlheim commented on .travis.yml:
```diff
@@ -235,6 +235,17 @@ matrix:
         - USE_SYSLIBS=false
         - SHARED_LIBSCSYNTH=false
 
+    # macOS 10.13 xcode10 legacy build with qt 5.9
```
this should probably replace the xcode10.3 build; we don't need both of these

brianlheim reviewed (approved):
looks good, thanks!

brianlheim commented on editors/sc-ide/CMakeLists.txt:
```diff
@@ -498,6 +498,8 @@ if(APPLE OR WIN32)
                 message(STATUS \"Fixing up bad path in libfftw3f.3.dylib manually due to faulty logic in macdeployqt\")
                 execute_process(COMMAND install_name_tool -change /usr/local/lib/gcc/0/libgcc_s.1.dylib
                     @loader_path/libgcc_s.1.dylib ${CONTENTS_DIR}/Frameworks/libfftw3f.3.dylib)
+                execute_process(COMMAND install_name_tool -change /usr/local/lib/gcc/9/libgcc_s.1.dylib
```
this was a typo -- you can delete the above line with `/0/`

brianlheim commented on .travis/before-install-osx.sh:
```diff
@@ -2,12 +2,22 @@
+    #run update first so that possible update errors won't hold up package installation
+    brew update
+else
+    export HOMEBREW_NO_AUTO_UPDATE=1
```
always export this

brianlheim commented on .travis/before-install-osx.sh:
```diff
@@ -2,12 +2,22 @@
-brew install libsndfile || brew install libsndfile || exit 1
+if $UPDATE_HOMEBREW; then
+    #run update first so that possible update errors won't hold up package installation
+    brew update
```
`brew update --preinstall` is faster and does the same thing

dyfer replied:
Also should I disable update for legacy build or leave it in?

brianlheim commented on .travis.yml:
```diff
@@ -235,32 +235,39 @@ matrix:
-        - JOB_NAME="macos xcode10.3"
+        - JOB_NAME="macos xcode10.1 legacy build with qt 5.9"
+        - UPDATE_HOMEBREW=false
+        - QT_FORMULA="supercollider/formulae/qt@5.9.3"
```
change UPDATE_HOMEBREW to IS_LEGACY_BUILD, remove QT_FORMULA, and just use the name of the formula directly in the script

dyfer replied:
This was intentional. My idea was that we would have control over updating homebrew in the configuration script, so we   could easily turn in on/off for individual builds if desired. Similarly, specifying QT formula in the script would make it possible to choose a separate formula for other builds - probably only in the situations like with Qt 5.15.1, but still I thought it would be nice to have this configurability.
If you still think it should be changed to a single variable and the formula being hardcoded in the script, I'm happy to change this.

brianlheim replied:
yes, please change it. this does not seem like a likely enough future that we need to make the existing code more complicated

brianlheim commented on .travis/before-install-osx.sh:
```diff
@@ -2,12 +2,22 @@
+    brew install $QT_FORMULA --force || exit 4
+elif $UPDATE_HOMEBREW; then
+    brew upgrade qt5 || exit 4
+fi
```
after the changes i note above this can become

```bash
if $IS_LEGACY_BUILD; then
  brew install <qt5.9-formula> --force || exit 4
else
  brew upgrade qt5 || exit 4
  brew link --force qt5 || exit 5
fi
```

dyfer replied:
Sounds good.

BTW I was dropping `brew link`. Do you know why is it here? AFAIU it would link qt to a directory in the PATH but I don't think this is needed. We get qt path from homebrew anyway through ``` -DCMAKE_PREFIX_PATH=`brew --prefix qt5` ```. I never needed to link qt to make SC build.

brianlheim replied:
oh good point, yeah it can be dropped. thanks!

brianlheim reviewed (changes requested):
looks good overall, let's get CI fixed now and then see how this works

brianlheim commented on .travis/before-install-osx.sh:
```diff
@@ -2,12 +2,22 @@
 export HOMEBREW_NO_ANALYTICS=1
 
-brew install libsndfile || brew install libsndfile || exit 1
+if $UPDATE_HOMEBREW; then
```
also add `export HOMEBREW_NO_INSTALL_CLEANUP=1` in all builds, this prevents homebrew from doing an expensive cleanup step rarely ("brew clean has not been run in X days, running now")

brianlheim commented:
i just realized this is deploying to github and S3. is that intended?

dyfer commented:
> i just realized this is deploying to github and S3. is that intended?

Yes... but I don't feel strongly about this. Should I leave github only?

dyfer commented:
I think this should be ready for another review now.

I made it deploy to Github but not AWS. However this logic is now split between the build matrix and the deploy step. Is that okay? or should I change it to maybe have two deploy flags, like `$DO_DEPLOY_AWS` and `$DO_DEPLOY_GITHUB` in the build matrix?

brianlheim commented:
sorry, i just realized my thumbs up reaction may have read as me saying "yes do this". i don't have a strong feeling either way, but it's probably good to deploy to S3 as well if we have the space. i don't know who actually controls/pays for it, we should ask them.

dyfer commented:
No worries, thanks for clarifying. Trying to make deployment independent prompted me to look into the package script and change some hardcoded paths there, I think the changes currently are still desirable.

BTW how to test deployment itself? for S3, I guess I could push this branch to the supercollider repo. As for GH - would we wait for the release and make any adjustments there? Or should I make a temporary tag to test this?

brianlheim commented on .travis.yml:
```diff
@@ -303,16 +310,17 @@ script:
- - if [[ ! -z "$DO_DEPLOY" ]]; then $TRAVIS_BUILD_DIR/.travis/package-$TRAVIS_OS_NAME.sh; fi
+ - echo $DEPLOY_AWS_PREFIX
+ - echo $DEPLOY_GITHUB_SUFFIX
+ - if [[ ! -z "$DEPLOY_AWS_PREFIX" ]] || [[ ! -z "$DEPLOY_GITHUB_SUFFIX" ]]; then $TRAVIS_BUILD_DIR/.travis/package-$TRAVIS_OS_NAME.sh; fi
```
simplify to `if [[ -n "..." || -n "..." ]]; then ...`

brianlheim commented on .travis.yml:
```diff
@@ -333,21 +341,21 @@ deploy:
 
 after_deploy:
- - "echo S3 Build Location: $S3_URL"
+ - 'if [[ ! -z "$DEPLOY_AWS_PREFIX" ]]; then echo S3 Build Location: $S3_URL; fi' 
```
don't think you need the single quotes? and can simplify to `-n` instead of `! -z`. you can do that throughout this script if you want in fact

dyfer replied:
I think I was getting errors without the quotes, so left them.

brianlheim commented on .travis/before-install-osx.sh:
```diff
@@ -1,13 +1,23 @@
+  brew install supercollider/formulae/qt@5.9.3 --force || exit 4
+else
+  brew upgrade qt5 || exit 4
+fi
```
indent at 4 spaces please

brianlheim commented on .travis.yml:
```diff
@@ -235,32 +235,38 @@ matrix:
         - SHARED_LIBSCSYNTH=false
-        - DO_DEPLOY=YES # presence option
+        - DEPLOY_AWS_PREFIX="SC-" #turn on AWS deployment
+        - DEPLOY_GITHUB_SUFFIX="-macOS" #turn on Github deployment
```
space between # and comment text throughout please

brianlheim commented on .travis.yml:
```diff
@@ -303,16 +310,17 @@ script:
  # required for github releases
  - export BUILD_PREFIX=$TRAVIS_REPO_SLUG/$TRAVIS_OS_NAME
+ - export ARTIFACT_AWS_NAME = $DEPLOY_AWS_PREFIX$TRAVIS_COMMIT
+ - export ARTIFACT_GH_NAME = $SuperCollider-$VERSION_TO_BUILD$DEPLOY_GITHUB_SUFFIX
```
i don't think that `$` before SuperCollider should be there? also, when putting several environment variables right next to each other, using `${ABC}` instead of `$ABC` can improve readability and provide a bit of future proofing. i think i've even seen the recommendation to always use this notation.

brianlheim replied:
also also, the names "S3_ARTIFACT_NAME" and "GITHUB_ARTIFACT_NAME" would stylistically match better with the naming throughout

brianlheim replied:
also match style by removing the spaces around `=`s

dyfer replied:
ah, good catch with `$SuperCollider`, thanks! also 👍  for all other comments

brianlheim reviewed (changes requested):
thanks so much!

> BTW how to test deployment itself? for S3, I guess I could push this branch to the supercollider repo. As for GH - would we wait for the release and make any adjustments then? Or should I make a temporary tag to test this?

i want to test this now because there are a lot of changes. once you've addressed this round of comments you can push to a branch on this repo and then push a tag to test. make sure the tag does not match the glob `Version-*`. i can do it as well if you'd rather not touch it yourself.

dyfer commented on .travis.yml:
```diff
@@ -303,16 +309,17 @@ script:
- - if [[ -z "$DO_DEPLOY" ]]; then echo hi; fi
- - if [[ ! -z "$DO_DEPLOY" ]]; then $TRAVIS_BUILD_DIR/.travis/package-$TRAVIS_OS_NAME.sh; fi
+ - echo $DEPLOY_AWS_PREFIX
+ - echo $DEPLOY_GITHUB_SUFFIX
```
yeah, I thought it might've been left accidentally. However, my point was that I'd personally like a little more verbosity with deployment, to help knowing what's happening. I understand the above might not be the way to go.
I was thinking about something along those lines:
```
will try to deploy / will not try to not deploy to AWS
will try to deploy / will not try to not deploy to Github
```
before the deploy step, and
```
deployed / not deployed to AWS
deployed / not deployed to Github
```
during/after, because deployment is conditional on the commit being in the SC repo, having a tag etc.

What do you think about adding these? is this too much?

brianlheim replied:
i think it's just not very helpful to see this information there. most of this is already logged and documented in other ways. for example "Skipping a deployment with the {s3,releases} provider because [reason]", which appears at the end of a branch commit build log in nice bright bold yellow text, is straightforward enough once you know that we deploy to s3 and github releases. these additional log lines would be hidden inside drop-downs in the interactive build log view, too, making them less discoverable.

btw, there are like 5 AWS services we can potentially deploy to through travis, so it's clearer to refer to it as s3 rather than AWS. the reason AWS_KEY and AWS_SECRET are named what they in our scripts is because they are credentials for an AWS account, and could be reused for other services.

dyfer commented:
I think I addressed all comments. I pushed to the main repo and added a [test tag](https://github.com/supercollider/supercollider/releases/tag/3.11-travis-test-01), let's see what happens...

dyfer commented:
WIP - working on deployment

dyfer commented:
I realized that I needed qt 5.15.1 fix for builds to pass. 
Legacy build doesn't need that and I've seen it pass, but it timed out when deploying to github... OTOH I can see its [artifact here](https://github.com/supercollider/supercollider/releases/tag/untagged-b4eca043a06106961d57).
I've now pushed tag [3.11-travis-test-06](https://github.com/supercollider/supercollider/releases/tag/3.11-travis-test-06), let's see if this works.

dyfer commented:
I'm filing an issue about building with XCode 11 separately: #5188

brianlheim commented on .travis/package-osx.sh:
```diff
@@ -3,5 +3,8 @@
+zip -q -r --symlinks $HOME/artifacts/$S3_ARTIFACT_NAME.zip SuperCollider
+ls -a $HOME/artifacts
+cp $HOME/artifacts/$S3_ARTIFACT_NAME.zip $HOME/$GITHUB_ARTIFACT_NAME.zip
+ls -a $HOME
```
did you leave these ls's here intentionally? i don't mind them, just wondering

dyfer replied:
nope! that's the only debug I overlooked. I kind of like this output. But I should remove that, right?

brianlheim replied:
maybe just comment it out, it could still be helpful for debugging in the future

dyfer replied:
Sorry, didn't see this on time before rebasing. Should I add these back?

brianlheim replied:
nah it's fine, thank you for noticing though

brianlheim reviewed (changes requested):
thanks! can you please rebase this on develop to pick up the qt 5.15.1 fix instead of cherry-picking the commit onto your branch? after that i think this is ready to merge. thanks so much for this, it's incredibly helpful; i know exactly how annoying it is to test these kinds of changes.

brianlheim commented:
thanks! i'll do the cherry pick
